bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerMetadataSerDe.java

```diff
@@ -51,68 +58,100 @@
+        checkState(LINE_SPLITTER_BYTES.length == 1, "LINE_SPLITTER must be single byte");
 
-        for (Map.Entry<Long, ? extends List<BookieSocketAddress>> entry : metadata.getAllEnsembles().entrySet()) {
-            s.append(LINE_SPLITTER).append(entry.getKey());
-            for (BookieSocketAddress addr : entry.getValue()) {
-                s.append(FIELD_SPLITTER);
-                s.append(addr.toString());
+        for (int i = 0; i < VERSION_KEY_BYTES.length; i++) {
+            int b = is.read();
+            if (b < 0 || ((byte) b) != VERSION_KEY_BYTES[i]) {
+                throw new IOException("Ledger metadata header corrupt at index " + i);
             }
         }
-
-        if (metadata.getState() == State.IN_RECOVERY) {
-            s.append(LINE_SPLITTER).append(V1_IN_RECOVERY_ENTRY_ID)
-                .append(FIELD_SPLITTER).append(V1_CLOSED_TAG);
-        } else if (metadata.getState() == State.CLOSED) {
-            s.append(LINE_SPLITTER).append(metadata.getLastEntryId())
-                .append(FIELD_SPLITTER).append(V1_CLOSED_TAG);
-        } else {
-            checkArgument(metadata.getState() == State.OPEN,
-                          String.format("Unknown state %s for V1 serialization", metadata.getState()));
+        byte[] versionBuf = new byte[MAX_VERSION_DIGITS];
+        int i = 0;
+        while (i < MAX_VERSION_DIGITS) {
+            int b = is.read();
+            if (b == LINE_SPLITTER_BYTES[0]) {
+                String versionStr = new String(versionBuf, 0, i, UTF_8);
+                try {
+                    return Integer.parseInt(versionStr);
+                } catch (NumberFormatException nfe) {
+                    throw new IOException("Unable to parse version number from " + versionStr);
+                }
+            } else if (b < 0) {
+                break;
+            } else {
+                versionBuf[i++] = (byte) b;
+            }
         }
+        throw new IOException("Unable to find end of version number, metadata appears corrupt");
+    }
 
+    public byte[] serialize(LedgerMetadata metadata) throws IOException {
+        int formatVersion = metadata.getMetadataFormatVersion();
+        final byte[] serialized;
+        switch (formatVersion) {
+        case METADATA_FORMAT_VERSION_3:
+            serialized = serializeVersion3(metadata);
+            break;
+        case METADATA_FORMAT_VERSION_2:
+            serialized = serializeVersion2(metadata);
+            break;
+        case METADATA_FORMAT_VERSION_1:
+            serialized = serializeVersion1(metadata);
+            break;
+        default:
+            throw new IllegalArgumentException("Invalid format version " + formatVersion);
+        }
         if (log.isDebugEnabled()) {
-            log.debug("Serialized config: {}", s);
+            log.debug("Serialized with format {}: {}", formatVersion, Base64.getEncoder().encodeToString(serialized));
```

Comment:
how does `encodeToString` work for v1 and v2 metadata? since v1 and v2 are already printable strings.

Reply:
Changed to print string if version <= VERSION_2